helix-tui/src/buffer.rs

```diff
@@ -301,7 +301,7 @@ impl Buffer {
         y: u16,
         string: S,
         width: usize,
-        style: Style,
+        style: impl Fn(usize) -> Style, // Map a grapheme's string offset to a style
```

Comment:
This is a significant hotpath and we'll also get some code bloat from monomorphization. Why not just do this in `render()` by calling `set_string` multiple times? I'm only comfortable with this change if the base case (fixed style) gets reasonably inlined by rust.

Reply:
> Why not just do this in `render()` by calling `set_string` multiple times?

Because then I'd have to implement truncating the string (and adding an ellipsis if necessary) manually.

Code bloat is limited, since `set_string_truncated` is only called three times. To ensure the hotpath doesn't get slower, I'd actually prefer replacing the usage in `set_stringn` (which always sets `ellipsis` and `truncate_start` to `false`) with a specialized function. This would be much simpler and might even be more efficient than the current implementation.

@archseer what do you think?

Reply:
ping @archseer

Reply:
I'll finish my review later this weekend

Reply:
I don't quite like the duplication but it's fine for now 👍🏻